Implemented a simpler way for samples to request features

ZandroFargnoli opened the pull request:
## Description

This PR provides a cleaner solution to GPU extension features.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

gary-sweet committed:
Adding Vulkan-style statistics counters where available

Previously, the 'stats' interface just used HWCPipe to
interrogate performance counters from the device. HWCPipe
is not widely supported, so this really only worked for
Arm devices.

This change allows multiple 'StatsProviders' to be used.
Each provider can supply all, some, or none of the requested
counters depending on its capabilities. Any counters not
supplied by one provider will be requested of the next in
the list until one is found, or we run out of providers.
So, CPU counters can come from HWCPipe for example, whilst
GPU counters come from Vulkan's extension which has been
added as another provider.

The StatsProvider defines an abstract interface that is
implemented by the concrete providers. The parameters used
for graphing the statistics have moved from the GUI classes
into the StatsProvider also. This allows providers to modify
the names, units and scaling of counters where these differ
from the default values.

gary-sweet committed:
Fixed up doxygen and copyright warnings

ZandroFargnoli committed:
Added panning camera to AFBC sample to help show delta (#102)

* Added inactive camera panning to AFBC sample to help show delta

* Amended tutorial

gary-sweet committed:
Hopefully fixed the Windows build

JoseEmilio-ARM committed:
Add support for Arm Mali G77 counters (#119)

ZandroFargnoli committed:
Removed warning from gltf loader (#123)

frguthmann committed:
Corrected minor typo (#124)

JoseEmilio-ARM committed:
Fix typo in Swapchain Images tutorial (#122)

gary-sweet committed:
Addressed some formatting and naming feedback

SaschaWillems committed:
Proper compute n-body sample synchronization (#75)

* Add acquire and release buffer barriers if compute and graphics queue family indices differ

DO NOT MERGE

* Updated access flags and pipeline stage ordering for compute acquire barrier

* Add initial transfer from graphics to compute queue and back

* Fix copyright and clang format

Co-authored-by: Brad Davis <bdavis@saintandreas.org>

gary-sweet committed:
Moved some things into physical device

Simplifies code in device

gary-sweet committed:
Fixed incorrect namespace comment

ZandroFargnoli committed:
Implemented a simpler way for samples to request features

ZandroFargnoli committed:
Descriptor set layout validation is more straight forward

ZandroFargnoli committed:
Removed template from doxygen

gary-sweet reviewed (changes requested):
All PRs should be done from a fork of the repo, not a local branch as previously discussed.

This change is going to heavily clash with the changes you requested in my stats PR.
Can this be rebased on that to avoid extra work for both myself and the testers?

gary-sweet committed:
Tweak to Broadcom specific counters

gary-sweet committed:
Reworked top level stats interface and construction

Based on private feedback from Zandro, I've moved the
construction of the stats object into the base
VulkanSamples class which leaves individual samples
to just request the samples they would like.

The active_frame_index is no longer passed to
update_stats() or the command_buffer_begun/ended()
methods. It is fetched internally by the providers
that require it.

gary-sweet committed:
Fix Windows build

size_t -> uint32_t

ZandroFargnoli committed:
Rebased on Garys changes

ZandroFargnoli committed:
Format
